chapter_recurrent-neural-networks/rnn-concise.ipynb

```diff
@@ -0,0 +1,808 @@
+    "\n",
+    "        Tracker lrt = Tracker.fixed(lr);\n",
+    "        Optimizer sgd = Optimizer.sgd().setLearningRateTracker(lrt).build();\n",
+    "\n",
+    "        DefaultTrainingConfig config =\n",
+    "                new DefaultTrainingConfig(loss)\n",
+    "                        .optOptimizer(sgd) // Optimizer (loss function)\n",
+    "                        .optInitializer(\n",
+    "                                new NormalInitializer(0.01f),\n",
+    "                                Parameter.Type.WEIGHT) // setting the initializer\n",
+    "                        .optDevices(Device.getDevices(1)) // setting the number of GPUs needed\n",
+    "                        .addEvaluator(new Accuracy()) // Model Accuracy\n",
+    "                        .addTrainingListeners(TrainingListener.Defaults.logging()); // Logging\n",
+    "\n",
+    "        Trainer trainer = model.newTrainer(config);\n",
+    "        updater = (batchSize, subManager) -> trainer.step();\n",
+    "    }\n",
+    "\n",
+    "    Function<String, String> predict =\n",
+    "            (prefix) -> predictCh8(prefix, 50, net, vocab, device, manager);\n",
+    "    // Train and predict\n",
+    "    double ppl = 0.0;\n",
+    "    double speed = 0.0;\n",
+    "    for (int epoch = 0; epoch < numEpochs; epoch++) {\n",
+    "        Pair<Double, Double> pair =\n",
+    "                trainEpochCh8(net, dataset, loss, updater, device, useRandomIter, manager);\n",
+    "        ppl = pair.getKey();\n",
+    "        speed = pair.getValue();\n",
+    "        if ((epoch + 1) % 10 == 0) {\n",
+    "           animator.add(epoch + 1, (float) ppl, \"ppl\");\n",
+    "           animator.show();\n",
+    "        }\n",
+    "    }\n",
+    "    System.out.format(\n",
+    "            \"perplexity: %.1f, %.1f tokens/sec on %s%n\", ppl, speed, device.toString());\n",
+    "    System.out.println(predict.apply(\"time traveller\"));\n",
+    "    System.out.println(predict.apply(\"traveller\"));\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "/** Train a model within one epoch. */\n",
+    "public static Pair<Double, Double> trainEpochCh8(\n",
+    "        Object net,\n",
+    "        RandomAccessDataset dataset,\n",
+    "        Loss loss,\n",
+    "        Functions.voidTwoFunction<Integer, NDManager> updater,\n",
+    "        Device device,\n",
+    "        boolean useRandomIter,\n",
+    "        NDManager manager)\n",
+    "        throws IOException, TranslateException {\n",
+    "    StopWatch watch = new StopWatch();\n",
+    "    watch.start();\n",
+    "    Accumulator metric = new Accumulator(2); // Sum of training loss, no. of tokens\n",
+    "\n",
+    "    try (NDManager childManager = manager.newSubManager()) {\n",
+    "        NDList state = null;\n",
+    "        for (Batch batch : dataset.getData(manager)) {\n",
```

Comment:
you can use dataset.getData(childManager). All the output data will be attached to childManager and you can get rid of X.attach(childManager) and Y.attach(childManager)

Reply:
Done